Sphinx still imports from "distutils", which is going away in Python ≥ 3.12

### Describe the bug

Python 3.10 officially deprecated the standard `distutils` package after standardization of [PEP 632 -- Deprecate distutils module](https://www.python.org/dev/peps/pep-0632). Python 3.12 is slated to entirely remove that package – and justifiably so. It's an ongoing dumpster fire that can't be put out soon enough. :fire: 

Sadly, Sphinx still widely imports from `distutils`. [GitHub shows at least twelve pending references](https://github.com/sphinx-doc/sphinx/search?q=distutils&type=code). As a *very* temporary circumvention, globally replacing `distutils` with `setuptools._distutils` should suffice to resolve this. Of course, that's also guaranteed to blow up. Like CPython, `setuptools` intends to disembowel and eventually remove `setuptools._distutils`. Ain't nobody got volunteer time to maintain cruft in perpetuity.

## Does This Really Matter *Now*?

**Yup.** I author @beartype, which [doesn't necessarily play well with Sphinx's `autodoc` extension (as detailed here)](https://github.com/sphinx-doc/sphinx/issues/9805). I've kludged around that incompatibility on my end with ad-hoc and inexplicably dangerous heuristics that conditionally disable @beartype when we vaguely *think* `autodoc` might be active. Nobody actually *knows* when `autodoc` is active, of course. There's currently no public Sphinx API to externally detect that. Cue #9805. `</ahem>`

Because that's dangerous, we [exercise that with a non-trivial functional test programmatically running `sphinx-build` from within the active `pytest` process by calling the `sphinx.cmd.build.main()` entry point](https://github.com/beartype/beartype/blob/main/beartype_test/a90_func/lib/test_sphinx.py). Works great – *under Python < 3.10*, that is. Under Python ≥ 3.10, Sphinx emits deprecation warnings all over the place like a flailing water buffalo besieged by leeches. For safety, our test suite treats any warning as a test failure. Chaos ensues with extreme `pytest` tracebacks resembling:

```
>       from sphinx.cmd.build import main as sphinx_build

get_test_func_data_lib_sphinx_dir = <function get_test_func_data_lib_sphinx_dir at 0x7f9579c875f0>
is_success = <function is_success at 0x7f9579c86db0>
tmp_path   = PosixPath('/tmp/pytest-of-leycec/pytest-29/test_sphinx0')

../../../beartype_test/a90_func/lib/test_sphinx.py:43: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
../lib/python3.10/site-packages/sphinx/cmd/build.py:25: in <module>
    from sphinx.application import Sphinx
        Any        = typing.Any
        IO         = <class 'typing.IO'>
        List       = typing.List
        SystemMessage = <class 'docutils.utils.SystemMessage'>
        __builtins__ = <builtins>
        __cached__ = '/home/leycec/py/beartype/.tox/py310/lib/python3.10/site-packages/sphinx/cmd/__pycache__/build.cpython-310.pyc'
        __display_version__ = '4.2.0'
        __doc__    = '\n    sphinx.cmd.build\n    ~~~~~~~~~~~~~~~~\n\n    Build documentation from a provided source.\n\n    :copyright: Copyright 2007-2021 by the Sphinx team, see AUTHORS.\n    :license: BSD, see LICENSE for details.\n'
        __file__   = '/home/leycec/py/beartype/.tox/py310/lib/python3.10/site-packages/sphinx/cmd/build.py'
        __loader__ = <_frozen_importlib_external.SourceFileLoader object at 0x7f9579a118b0>
        __name__   = 'sphinx.cmd.build'
        __package__ = 'sphinx.cmd'
        __spec__   = ModuleSpec(name='sphinx.cmd.build', loader=<_frozen_importlib_external.SourceFileLoader object at 0x7f9579a118b0>, origin='/home/leycec/py/beartype/.tox/py310/lib/python3.10/site-packages/sphinx/cmd/build.py')
        argparse   = <module 'argparse' from '/usr/lib/python3.10/argparse.py'>
        bdb        = <module 'bdb' from '/usr/lib/python3.10/bdb.py'>
        locale     = <module 'locale' from '/usr/lib/python3.10/locale.py'>
        multiprocessing = <module 'multiprocessing' from '/usr/lib/python3.10/multiprocessing/__init__.py'>
        os         = <module 'os' from '/usr/lib/python3.10/os.py'>
        package_dir = '/home/leycec/py/beartype/.tox/py310/lib/python3.10/site-packages/sphinx'
        pdb        = <module 'pdb' from '/usr/lib/python3.10/pdb.py'>
        sphinx     = <module 'sphinx' from '/home/leycec/py/beartype/.tox/py310/lib/python3.10/site-packages/sphinx/__init__.py'>
        sys        = <module 'sys' (built-in)>
        traceback  = <module 'traceback' from '/usr/lib/python3.10/traceback.py'>
../lib/python3.10/site-packages/sphinx/application.py:34: in <module>
    from sphinx.domains import Domain, Index
        Any        = typing.Any
        Callable   = typing.Callable
        Config     = <class 'sphinx.config.Config'>
        Dict       = typing.Dict
        Directive  = <class 'docutils.parsers.rst.Directive'>
        Element    = <class 'docutils.nodes.Element'>
        IO         = <class 'typing.IO'>
        Lexer      = <class 'pygments.lexer.Lexer'>
        List       = typing.List
        Optional   = typing.Optional
        Parser     = <class 'docutils.parsers.Parser'>
        RemovedInSphinx60Warning = <class 'sphinx.deprecation.RemovedInSphinx60Warning'>
        StringIO   = <class '_io.StringIO'>
        TYPE_CHECKING = False
        TextElement = <class 'docutils.nodes.TextElement'>
        Transform  = <class 'docutils.transforms.Transform'>
        Tuple      = typing.Tuple
        Type       = typing.Type
        Union      = typing.Union
        __builtins__ = <builtins>
        __cached__ = '/home/leycec/py/beartype/.tox/py310/lib/python3.10/site-packages/sphinx/__pycache__/application.cpython-310.pyc'
        __doc__    = '\n    sphinx.application\n    ~~~~~~~~~~~~~~~~~~\n\n    Sphinx application class and extensibility interface.\n\n    ...n\n    :copyright: Copyright 2007-2021 by the Sphinx team, see AUTHORS.\n    :license: BSD, see LICENSE for details.\n'
        __file__   = '/home/leycec/py/beartype/.tox/py310/lib/python3.10/site-packages/sphinx/application.py'
        __loader__ = <_frozen_importlib_external.SourceFileLoader object at 0x7f9579a3e8f0>
        __name__   = 'sphinx.application'
        __package__ = 'sphinx'
        __spec__   = ModuleSpec(name='sphinx.application', loader=<_frozen_importlib_external.SourceFileLoader object at 0x7f9579a3e8f0>, origin='/home/leycec/py/beartype/.tox/py310/lib/python3.10/site-packages/sphinx/application.py')
        deque      = <class 'collections.deque'>
        locale     = <module 'sphinx.locale' from '/home/leycec/py/beartype/.tox/py310/lib/python3.10/site-packages/sphinx/locale/__init__.py'>
        nodes      = <module 'docutils.nodes' from '/home/leycec/py/beartype/.tox/py310/lib/python3.10/site-packages/docutils/nodes.py'>
        os         = <module 'os' from '/usr/lib/python3.10/os.py'>
        package_dir = '/home/leycec/py/beartype/.tox/py310/lib/python3.10/site-packages/sphinx'
        path       = <module 'posixpath' from '/usr/lib/python3.10/posixpath.py'>
        pickle     = <module 'pickle' from '/usr/lib/python3.10/pickle.py'>
        platform   = <module 'platform' from '/usr/lib/python3.10/platform.py'>
        roles      = <module 'docutils.parsers.rst.roles' from '/home/leycec/py/beartype/.tox/py310/lib/python3.10/site-packages/docutils/parsers/rst/roles.py'>
        sphinx     = <module 'sphinx' from '/home/leycec/py/beartype/.tox/py310/lib/python3.10/site-packages/sphinx/__init__.py'>
        sys        = <module 'sys' (built-in)>
        warnings   = <module 'warnings' from '/usr/lib/python3.10/warnings.py'>
../lib/python3.10/site-packages/sphinx/domains/__init__.py:24: in <module>
    from sphinx.roles import XRefRole
        ABC        = <class 'abc.ABC'>
        Any        = typing.Any
        Callable   = typing.Callable
        Dict       = typing.Dict
        Element    = <class 'docutils.nodes.Element'>
        Inliner    = <class 'docutils.parsers.rst.states.Inliner'>
        Iterable   = typing.Iterable
        List       = typing.List
        NamedTuple = <function NamedTuple at 0x7f9582bbda70>
        Node       = <class 'docutils.nodes.Node'>
        Optional   = typing.Optional
        SphinxError = <class 'sphinx.errors.SphinxError'>
        TYPE_CHECKING = False
        Tuple      = typing.Tuple
        Type       = typing.Type
        Union      = typing.Union
        _          = <function get_translation.<locals>.gettext at 0x7f9579a49180>
        __builtins__ = <builtins>
        __cached__ = '/home/leycec/py/beartype/.tox/py310/lib/python3.10/site-packages/sphinx/domains/__pycache__/__init__.cpython-310.pyc'
        __doc__    = '\n    sphinx.domains\n    ~~~~~~~~~~~~~~\n\n    Support for domains, which are groupings of description directives\n ...n\n    :copyright: Copyright 2007-2021 by the Sphinx team, see AUTHORS.\n    :license: BSD, see LICENSE for details.\n'
        __file__   = '/home/leycec/py/beartype/.tox/py310/lib/python3.10/site-packages/sphinx/domains/__init__.py'
        __loader__ = <_frozen_importlib_external.SourceFileLoader object at 0x7f95790f8370>
        __name__   = 'sphinx.domains'
        __package__ = 'sphinx.domains'
        __path__   = ['/home/leycec/py/beartype/.tox/py310/lib/python3.10/site-packages/sphinx/domains']
        __spec__   = ModuleSpec(name='sphinx.domains', loader=<_frozen_importlib_external.SourceFileLoader object at 0x7f95790f8370>, origi...__.py', submodule_search_locations=['/home/leycec/py/beartype/.tox/py310/lib/python3.10/site-packages/sphinx/domains'])
        abstractmethod = <function abstractmethod at 0x7f9582e0b3e0>
        cast       = <function cast at 0x7f9582bbc7e0>
        copy       = <module 'copy' from '/usr/lib/python3.10/copy.py'>
        nodes      = <module 'docutils.nodes' from '/home/leycec/py/beartype/.tox/py310/lib/python3.10/site-packages/docutils/nodes.py'>
        pending_xref = <class 'sphinx.addnodes.pending_xref'>
        system_message = <class 'docutils.nodes.system_message'>
../lib/python3.10/site-packages/sphinx/roles.py:20: in <module>
    from sphinx.util.docutils import ReferenceRole, SphinxRole
        Any        = typing.Any
        Dict       = typing.Dict
        Element    = <class 'docutils.nodes.Element'>
        List       = typing.List
        Node       = <class 'docutils.nodes.Node'>
        TYPE_CHECKING = False
        TextElement = <class 'docutils.nodes.TextElement'>
        Tuple      = typing.Tuple
        Type       = typing.Type
        _          = <function get_translation.<locals>.gettext at 0x7f9579a49180>
        __annotations__ = {}
        __builtins__ = <builtins>
        __cached__ = '/home/leycec/py/beartype/.tox/py310/lib/python3.10/site-packages/sphinx/__pycache__/roles.cpython-310.pyc'
        __doc__    = '\n    sphinx.roles\n    ~~~~~~~~~~~~\n\n    Handlers for additional ReST roles.\n\n    :copyright: Copyright 2007-2021 by the Sphinx team, see AUTHORS.\n    :license: BSD, see LICENSE for details.\n'
        __file__   = '/home/leycec/py/beartype/.tox/py310/lib/python3.10/site-packages/sphinx/roles.py'
        __loader__ = <_frozen_importlib_external.SourceFileLoader object at 0x7f95790f8960>
        __name__   = 'sphinx.roles'
        __package__ = 'sphinx'
        __spec__   = ModuleSpec(name='sphinx.roles', loader=<_frozen_importlib_external.SourceFileLoader object at 0x7f95790f8960>, origin='/home/leycec/py/beartype/.tox/py310/lib/python3.10/site-packages/sphinx/roles.py')
        addnodes   = <module 'sphinx.addnodes' from '/home/leycec/py/beartype/.tox/py310/lib/python3.10/site-packages/sphinx/addnodes.py'>
        nodes      = <module 'docutils.nodes' from '/home/leycec/py/beartype/.tox/py310/lib/python3.10/site-packages/docutils/nodes.py'>
        re         = <module 're' from '/usr/lib/python3.10/re.py'>
        system_message = <class 'docutils.nodes.system_message'>
        utils      = <module 'docutils.utils' from '/home/leycec/py/beartype/.tox/py310/lib/python3.10/site-packages/docutils/utils/__init__.py'>
        ws_re      = re.compile('\\s+')
../lib/python3.10/site-packages/sphinx/util/docutils.py:15: in <module>
    from distutils.version import LooseVersion
        __annotations__ = {}
        __builtins__ = <builtins>
        __cached__ = '/home/leycec/py/beartype/.tox/py310/lib/python3.10/site-packages/sphinx/util/__pycache__/docutils.cpython-310.pyc'
        __doc__    = '\n    sphinx.util.docutils\n    ~~~~~~~~~~~~~~~~~~~~\n\n    Utility functions for docutils.\n\n    :copyright: Copyright 2007-2021 by the Sphinx team, see AUTHORS.\n    :license: BSD, see LICENSE for details.\n'
        __file__   = '/home/leycec/py/beartype/.tox/py310/lib/python3.10/site-packages/sphinx/util/docutils.py'
        __loader__ = <_frozen_importlib_external.SourceFileLoader object at 0x7f95790f97c0>
        __name__   = 'sphinx.util.docutils'
        __package__ = 'sphinx.util'
        __spec__   = ModuleSpec(name='sphinx.util.docutils', loader=<_frozen_importlib_external.SourceFileLoader object at 0x7f95790f97c0>, origin='/home/leycec/py/beartype/.tox/py310/lib/python3.10/site-packages/sphinx/util/docutils.py')
        __warningregistry__ = {'version': 678}
        contextmanager = <function contextmanager at 0x7f9582d98470>
        copy       = <function copy at 0x7f9582c855a0>
        os         = <module 'os' from '/usr/lib/python3.10/os.py'>
        re         = <module 're' from '/usr/lib/python3.10/re.py'>
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

    """distutils
    
    The main package for the Python Module Distribution Utilities.  Normally
    used from a setup script as
    
       from distutils.core import setup
    
       setup (...)
    """
    
    import sys
    import warnings
    
    __version__ = sys.version[:sys.version.index(' ')]
    
    _DEPRECATION_MESSAGE = ("The distutils package is deprecated and slated for "
                            "removal in Python 3.12. Use setuptools or check "
                            "PEP 632 for potential alternatives")
>   warnings.warn(_DEPRECATION_MESSAGE,
                  DeprecationWarning, 2)
E   DeprecationWarning: The distutils package is deprecated and slated for removal in Python 3.12. Use setuptools or check PEP 632 for potential alternatives

_DEPRECATION_MESSAGE = 'The distutils package is deprecated and slated for removal in Python 3.12. Use setuptools or check PEP 632 for potential alternatives'
__builtins__ = <builtins>
__cached__ = '/usr/lib/python3.10/distutils/__pycache__/__init__.cpython-310.pyc'
__doc__    = 'distutils\n\nThe main package for the Python Module Distribution Utilities.  Normally\nused from a setup script as\n\n   from distutils.core import setup\n\n   setup (...)\n'
__file__   = '/usr/lib/python3.10/distutils/__init__.py'
__loader__ = <_frozen_importlib_external.SourceFileLoader object at 0x7f95790fa5d0>
__name__   = 'distutils'
__package__ = 'distutils'
__path__   = ['/usr/lib/python3.10/distutils']
__spec__   = ModuleSpec(name='distutils', loader=<_frozen_importlib_external.SourceFileLoader object at 0x7f95790fa5d0>, origin='/usr/lib/python3.10/distutils/__init__.py', submodule_search_locations=['/usr/lib/python3.10/distutils'])
__version__ = '3.10.0'
sys        = <module 'sys' (built-in)>
warnings   = <module 'warnings' from '/usr/lib/python3.10/warnings.py'>

/usr/lib/python3.10/distutils/__init__.py:19: DeprecationWarning
=================================================== short test summary info ===================================================
SKIPPED [1] ../../../beartype_test/a00_unit/a10_pep/test_pep563.py:227: Python 3.10.0 >= 3.10.0.
FAILED ../../../beartype_test/a90_func/lib/test_sphinx.py::test_sphinx - DeprecationWarning: The distutils package is deprec...
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! stopping after 1 failures !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
========================================== 1 failed, 183 passed, 1 skipped in 7.03s ===========================================
ERROR: InvocationError for command /home/leycec/py/beartype/.tox/py310/bin/python -W 'ignore:The distutils package is deprecated:DeprecationWarning' -m pytest --maxfail=1 /home/leycec/py/beartype (exited with code 1)
___________________________________________________________ summary ___________________________________________________________
ERROR:   py310: commands failed
```

To preserve our sanity, we're currently just ignoring that test under Python ≥ 3.10. We're not proud – but sometimes you just gotta get to sleep.

## Thanks for all the Structured Docos

As always, thanks a well-documented ton for all the *amazing* volunteer work everyone does here. Sphinx rocks! May this issue find the final reST it so deserves. :wink:

### How to Reproduce

Just look at the Sphinx codebase, maybe? Not sure what to say here. Badness is bad. The bitrot has gotta be removed – ideally sooner than later.

### Expected behavior

*...that Sphinx not import from `distutils`.* Work with me here, people.

### Your project

The problem is in Sphinx itself. Ain't no small project gonna help.

### Screenshots

Let's be honest: nobody wants another screenshot of my puce monochrome CLI ViM setup. *Nobody.*

### OS

Gentoo Linux, of course! O_o

### Python version

3.10

### Sphinx version

4.2.0

### Sphinx extensions

Irrelevant.

### Extra tools

Irrelevant.

### Additional context

Irrelevant.

--- a/sphinx/setup_command.py
+++ b/sphinx/setup_command.py
@@ -6,8 +6,6 @@
 import os
 import sys
 import warnings
-from distutils.cmd import Command
-from distutils.errors import DistutilsExecError
 from io import StringIO
 from typing import Any, Dict
 
@@ -17,6 +15,13 @@ from sphinx.deprecation import RemovedInSphinx70Warning
 from sphinx.util.console import color_terminal, nocolor
 from sphinx.util.docutils import docutils_namespace, patch_docutils
 from sphinx.util.osutil import abspath
+
+try:
+    from setuptools import Command
+    from setuptools.errors import ExecError
+except ImportError:
+    from distutils.cmd import Command
+    from distutils.errors import DistutilsExecError as ExecError
 
 
 class BuildDoc(Command):
@@ -97,7 +102,7 @@ class BuildDoc(Command):
         self.link_index = False
         self.copyright = ''
         # Link verbosity to distutils' (which uses 1 by default).
-        self.verbosity = self.distribution.verbose - 1  # type: ignore
+        self.verbosity = self.distribution.verbose - 1
         self.traceback = False
         self.nitpicky = False
         self.keep_going = False
@@ -125,7 +130,7 @@ class BuildDoc(Command):
 
         if self.build_dir is None:
             build = self.get_finalized_command('build')
-            self.build_dir = os.path.join(abspath(build.build_base), 'sphinx')  # type: ignore
+            self.build_dir = os.path.join(abspath(build.build_base), 'sphinx')
 
         self.doctree_dir = os.path.join(self.build_dir, 'doctrees')
 
@@ -139,7 +144,7 @@ class BuildDoc(Command):
 
         if not color_terminal():
             nocolor()
-        if not self.verbose:  # type: ignore
+        if not self.verbose:
             status_stream = StringIO()
         else:
             status_stream = sys.stdout  # type: ignore
@@ -171,8 +176,7 @@ class BuildDoc(Command):
                                  verbosity=self.verbosity, keep_going=self.keep_going)
                     app.build(force_all=self.all_files)
                     if app.statuscode:
-                        raise DistutilsExecError(
-                            'caused by %s builder.' % app.builder.name)
+                        raise ExecError('caused by %s builder.' % app.builder.name)
             except Exception as exc:
                 handle_exception(app, self, exc, sys.stderr)
                 if not self.pdb:

